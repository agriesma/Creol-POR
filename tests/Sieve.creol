interface Sieve
begin
with Any
op send(in x: Nat; out r: Bool)
end

class Generator
begin
  var n: Integer
  var next: Sieve
  op init == next := new Sieve(2); n := 3
  op run == var r: Bool; next.send(n; r); n := n + 1
end

class Sieve(p: Integer) implements Sieve
begin
  var next: Sieve
  op send(in n: Integer; out res: Bool) ==
    var d: Integer := n / p, r: Integer := n - d * p;
    if r = 0
    then res := false
    else if next /= nil
         then next.send(n, res)
         else next := new Sieve(n); res := true
         fi
    fi
end
