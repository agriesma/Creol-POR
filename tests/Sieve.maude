load creol-interpreter
mod PROGRAM is
pr CREOL-INTERPRETER .
op init : -> Configuration [ctor] .
eq init =
< "Generator" : Cl | Inh: noInh, Par: noVid, Att: "n" |-> null , "next" |-> null, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: "next" ::= new "Sieve" ( int(2) ) ; "n" ::= int(3) ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: "r" |-> null, Code: "sync.0" ! "next" . "send" ( "n" ) ; ( "sync.0" ? ( "r" ) )  ; "n" ::= "+" ( "n" # int(1) ) ; ( ".anon" ! "this" . "run" (emp ) )  ; free( ".anon" ) ; return ( emp ) >, Ocnt: 0 >

< "Sieve" : Cl | Inh: noInh, Par: "p", Att: "p" |-> null , "next" |-> null, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "send" : Mtdname | Param: "n", Latt: "n" |-> null , "res" |-> null , "d" |-> null , "r" |-> null, Code: "d" ::= "/" ( "n" # "p" ) ; "r" ::= "-" ( "n" # "*" ( "d" # "p" ) ) ; if "=" ( "r" # int(0) ) th "res" ::= bool(false) el if "/=" ( "next" # null ) th "sync.0" ! "next" . "send" ( "n" ) ; ( "sync.0" ? ( "res" ) )  el "next" ::= new "Sieve" ( "n" ) ; "res" ::= bool(true) fi fi ; return ( "res" ) >, Ocnt: 0 >

.
endm
