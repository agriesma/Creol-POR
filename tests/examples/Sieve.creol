/*
 * Sieve.creol -- Sieve of Erathostenes.
 *
 * This file is part of creoltools
 *
 * Written and Copyright (c) 2007 by Marcel Kyas
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

interface Sieve
begin
with Any
  op send(in x: Int; out r: Bool)
end

class Generator
begin
  var n: Int
  var next: Sieve
  op init == next := new Sieve(2); n := 3
  op run == var r: Bool; next.send(n; r); n := n + 1; !run()
end

class Sieve(p: Int) implements Sieve
begin
  var next: Sieve

  with Any
    op send(in n: Int; out res: Bool) ==
      var d: Int := n / p, r: Int := n - d * p;
      if r = 0
      then res := false
      else
	if next /= null
        then next.send(n; res)
        else next := new Sieve(n); res := true
        end
      end
end
