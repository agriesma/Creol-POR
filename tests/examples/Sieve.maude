load creol-interpreter
mod PROGRAM is
protecting CREOL-SIMULATOR .
op init : -> Configuration [ctor] .
eq init =
< "Generator" : Cl | Inh: noInh, Par: noVid, Att: "n" |-> null , "next" |-> null, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: "next" ::= new "Sieve" ( int(2) ) ; "n" ::= int(3) ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: "label:1" |-> null , "label:0" |-> null , "r" |-> null , "_" |-> null, Code: "label:0" ! "next" . "send" ( "n" ) ; ( "label:0" ? ( "r" ) )  ; "n" ::= "+" ( "n" :: int(1) ) ; ( "label:1" ! "this" . "run" (emp ) )  ; free( "label:1" ) ; return ( emp ) >, Ocnt: 0 >

< "Sieve" : Cl | Inh: noInh, Par: "p", Att: "p" |-> null , "next" |-> null, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return ( emp ) > *
  < "send" : Mtdname | Param: "n", Latt: "n" |-> null , "res" |-> null , "label:0" |-> null , "d" |-> null , "r" |-> null , "_" |-> null, Code: "d" ::= "/" ( "n" :: "p" ) ; "r" ::= "-" ( "n" :: "*" ( "d" :: "p" ) ) ; if "=" ( "r" :: int(0) ) th "res" ::= bool(false) el if "/=" ( "next" :: null ) th "label:0" ! "next" . "send" ( "n" ) ; ( "label:0" ? ( "res" ) )  el "next" ::= new "Sieve" ( "n" ) ; "res" ::= bool(true) fi fi ; return ( "res" ) >, Ocnt: 0 >

.
endm
