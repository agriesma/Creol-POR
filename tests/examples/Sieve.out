==========================================
reduce in PROGRAM : init .
result Configuration: 
< "Generator" : Cl | Inh: noInh, Par: noVid, Att: "n" |-> null, "next" |->
    null, Mtds:
  < "init" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: "next" ::= new
    "Sieve"(int(2)) ; "n" ::= int(3) ; return(emp) > *
  < "run" : Mtdname | Param: noVid, Latt: "_" |-> null, "label:0" |-> null,
    "label:1" |-> null, "r" |-> null, Code: "label:0" ! "next" . "send"("n") ;
    "label:0" ?("r") ; "n" ::= "+"("n" :: int(1)) ; "label:1" ! "this" . "run"(
    emp) ; free("label:1") ; return(emp) >,
  Ocnt: 0 >

< "Sieve" : Cl | Inh: noInh, Par: "p", Att: "next" |-> null, "p" |-> null,
    Mtds:
  < "init" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return(
    emp) > *
  < "run" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return(
    emp) > *
  < "send" : Mtdname | Param: "n", Latt: "_" |-> null, "d" |-> null, "label:0"
    |-> null, "n" |-> null, "r" |-> null, "res" |-> null, Code: "d" ::= "/"("n"
    :: "p") ; "r" ::= "-"("n" :: "*"("d" :: "p")) ; if "="("r" :: int(0)) th
    "res" ::= bool(false) el if "/="("next" :: null) th "label:0" ! "next" .
    "send"("n") ; "label:0" ?("res") el "next" ::= new "Sieve"("n") ; "res" ::=
    bool(true) fi fi ; return("res") >,
  Ocnt: 0 >
Bye.
